lib/livebook/file_system/s3/client.ex

```diff
@@ -0,0 +1,355 @@
+defmodule Livebook.FileSystem.S3.Client do
+  @moduledoc false
+
+  alias Livebook.FileSystem
+  alias Livebook.FileSystem.S3
+
+  @doc """
+  Sends a request to the bucket to get list of objects.
+  """
+  @spec list_objects(S3.t(), keyword()) :: {:ok, map()} | {:error, String.t()}
+  def list_objects(file_system, opts) do
+    prefix = opts[:prefix]
+    delimiter = opts[:delimiter]
+    query = %{"list-type" => "2", "prefix" => prefix, "delimiter" => delimiter}
+
+    map_bucket_keys = fn result, xml_key, map_key ->
+      items =
+        case result do
+          %{^xml_key => item} when is_map(item) -> [item]
+          %{^xml_key => items} when is_list(items) -> items
+          _ -> []
+        end
```

Comment:
I would still refactor into `xml_get_list` rather than anonymous function, since it's more general.